File: leetcode/141.环形链表/141.环形链表.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        fast = slow = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        return False

File: hot/leetcode/141.环形链表/141.环形链表.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}
/*
快慢指针
快指针二倍速，慢指针一倍速
如果有环，快指针追（fast==slow)上慢指针
*/
func hasCycle(head *ListNode) bool {
	fast := head
	slow := head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
        if fast == slow {
            return true
        }
	}
	return false
}
